### PR TITLE
[megatron, model] feat: add MegatronEngine, MegatronEngineForCausalLM

### PR DESCRIPTION
**Pull request by vermouth1992**
### What does this PR do?

- Add MegatronEngine, MegatronEngineForCausalLM
- Add a tests to show that Qwen2.5 0.5b instruct forward with megatron tp=2, cp=2, pp=2 matches with huggingface transformers model definition on a single machine
- Add both ppo loss and sft loss to show how to use MegatronEngine with different losses

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [ ] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `megatron`, `sglang`, `vllm`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `hardware`, `deployment`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`
  - If this PR involves multiple modules, separate them with `,` like `[megatron, fsdp, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][fsdp, megatron] feat: dynamic batching`

### Test

> For changes that can not be tested by CI (e.g., algorithm implementation, new model support), validate by experiment(s) and show results like training curve plots, evaluation results, etc.

### API and Usage Example

> Demonstrate how the API changes if any, and provide usage example(s) if possible.

```python
# Add code snippet or script demonstrating how to use this
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/volcengine/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/volcengine/verl/tree/main/docs).
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/volcengine/verl/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...
- [ ] Once your PR is ready for CI, send a message in [the `ci-request` channel](https://verl-project.slack.com/archives/C091TCESWB1) in [the `verl` Slack workspace](https://join.slack.com/t/verl-project/shared_invite/zt-3855yhg8g-CTkqXu~hKojPCmo7k_yXTQ). (If not accessible, please try [the Feishu group (飞书群)](https://applink.larkoffice.com/client/chat/chatter/add_by_link?link_token=772jd4f1-cd91-441e-a820-498c6614126a).)

### CODE REVIEW

File: verl/workers/engine/megatron/engine_impl.py [hunk 1]
```diff
@@ -163,4 +340,307 @@ def load_checkpoint(self, local_path, hdfs_path=None, del_local_after_load=True)
...
+                    "log_probs": log_probs,
+                }
+                if calculate_entropy:
+                    output["entropy"] = entropys
+                return output
+            else:
+                return {}
+
+
+class MegatronEngineForTokenClassification(MegatronEngine):
```
Comment (ETOgaosion) [inline verl/workers/engine/megatron/engine_impl.py]: Not fully implemented right? Will you extract similar operations in forward_backward batch here?

File: verl/workers/roles/losses.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
+# Copyright 2025 Bytedance Ltd. and/or its affiliates
```
Comment (ETOgaosion) [inline verl/workers/roles/losses.py]: Seems losses is not a role, maybe in folder like `roles/utils` is better?

File: verl/workers/engine/megatron/engine_impl.py [hunk 3]
```diff
@@ -163,4 +340,307 @@ def load_checkpoint(self, local_path, hdfs_path=None, del_local_after_load=True)
...
+        for module in self.engine.module:
+            module.eval()
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.engine._is_offload_param:
+            offload_megatron_model_to_cpu(self.engine.module)
+        self.engine.mode = None
+
+
+class EngineTrainModeCtx:
```
Comment (ETOgaosion) [inline verl/workers/engine/megatron/engine_impl.py]: Maybe this and `EngineEvalModeCtx` are common contexts between megatron and FSDP, shall we use a parent ctx class in `engine/base`?

File: verl/workers/engine/base.py [hunk 4]
```diff
@@ -125,28 +77,52 @@ def lr_scheduler_step(self):
         """
         raise NotImplementedError
 
-    def shard_data(self, data):
+    def forward_backward_batch(self, data: DataProto, loss_function: Callable, forward_only=False) -> Any:
```
Comment (ETOgaosion) [inline verl/workers/engine/base.py:80]: Would it be better to clarify which batch (seems a mini batch?) here and below train_batch?

File: verl/workers/engine/megatron/engine_impl.py (hunk 1 above)
Comment (vermouth1992) [reply verl/workers/engine/megatron/engine_impl.py]: Will probably write a new forward_backward_batch. Note quite sure we should combine them because there are some differences:
1. LMHead requires temperature input and fused kernels that are hard to combine with ValueHead

File: verl/workers/roles/losses.py (hunk 2 above)
Comment (vermouth1992) [reply verl/workers/roles/losses.py]: Sure. Let me fix this

File: verl/workers/engine/megatron/engine_impl.py [hunk 5]
```diff
@@ -163,4 +340,307 @@ def load_checkpoint(self, local_path, hdfs_path=None, del_local_after_load=True)
...
+    def __exit__(self, exc_type, exc_value, traceback):
+        if self.engine._is_offload_param:
+            offload_megatron_model_to_cpu(self.engine.module)
+        if self.engine._is_offload_optimizer:
+            offload_megatron_optimizer(optimizer=self.engine.optimizer)
+        self.engine.mode = None
+
+
+class MegatronEngineForCausalLM(MegatronEngine):
+    def forward_backward_batch(self, data: DataProto, loss_function: Callable, forward_only=False) -> Any:
```
Comment (ETOgaosion) [inline verl/workers/engine/megatron/engine_impl.py]: Seems logic here still complex, maybe there are ways to decrease the branches and refactor from design.

File: verl/workers/engine/megatron/engine_impl.py (hunk 3 above)
Comment (vermouth1992) [reply verl/workers/engine/megatron/engine_impl.py]: There are some differences. In FSDP, the module is FSDPModule while in megatron, it's a list of model chunk. The API for load/offload is also different.

File: verl/workers/engine/base.py (hunk 4 above)
Comment (vermouth1992) [reply verl/workers/engine/base.py:80]: `train_batch` is a base class method that does
- optimizer.zero_grad
- forward_backward_batch
- optimizer.step

The batch will be directly passed in. Will need to add documentation to show what should be inside batch

File: verl/workers/engine/base.py [hunk 6]
```diff
@@ -125,28 +77,52 @@ def lr_scheduler_step(self):
...
         """
-        Shard or partition data for distributed training or parallel execution.
+        Perform a forward pass and optionally a backward pass on a batch of data.
 
         Args:
-            data: Data structure to be sharded across devices/workers.
+            data: The input data for the forward pass, typically containing tensors and metadata.
```
Comment (ccclyu) [inline verl/workers/engine/base.py:85]: nit: miss one args `loss_function`

File: verl/workers/engine/megatron/engine_impl.py (hunk 5 above)
Comment (ccclyu) [reply verl/workers/engine/megatron/engine_impl.py]: same feeling as @ETOgaosion and many if-else branches will make code management difficult.

**Comment by vermouth1992** [comment]
I have made the code more modular, please review again. Thanks!

File: verl/workers/engine/base.py (hunk 6 above)
Comment (vermouth1992) [reply verl/workers/engine/base.py:85]: Fixed

File: verl/workers/engine/base.py (hunk 4 above)
Comment (vermouth1992) [reply verl/workers/engine/base.py:80]: There is no notion of "mini-batch" for engine. We will use global batch. mini-batch is a notion in PPO training, where we split the batch from collections and split into several mini-batches

File: verl/workers/engine/megatron/engine_impl.py (hunk 5 above)
Comment (vermouth1992) [reply verl/workers/engine/megatron/engine_impl.py]: I have made the code modular and avoid inner functions to make the interface clean

**Comment by wwwjn** [comment]
Hi @vermouth1992 a quick question: IIUC This PR implemented Megatron engine as an Actor, did you plug/tested it in PPO? Thank you!

File: verl/workers/roles/utils/losses.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+
+
+def sft_loss(config: ActorConfig, model_output, data):
+    log_prob = model_output["log_probs"]  # [bsz, response_length]
+    response_mask = data["response_mask"].to(bool)
+    loss = -torch.mean(log_prob * response_mask)
+    return loss, {"loss": loss.detach().item()}
+
+
+def ppo_loss(config: ActorConfig, model_output, data):
```
Comment (ISEEKYAN) [inline verl/workers/roles/utils/losses.py:29]: this function need info of global_batch_size or global_number_tokens to correctly calculate loss_scale_factor, for correct gradient accumulation
